README.md
# flinit <kbd>0.2-beta</kbd>

Flinit ( **fl**ask **init** ) is a  Python Command-Line Tool to Initialize a Python-Flask Project with Boilerplate Code and other required tools.

Flinit is a create-react-app like tool for Flask Projects.

## Features
- Creates Project Working Directory
- Spins up a Python Virtual Environment
- Generates Flask App File, <kbd>app.py</kbd>
- Installs Flask and CORS (optional)
- Adds Flask and CORS Helper Methods
- Adds Boilerplate Code to <kbd>app.py</kbd>
- Generates initial environment requirments file
- Adds <kbd>README.md</kbd> to the project
- Generates <kbd>.gitignore</kbd> for Python Project
- Initializes Git for Project.


# Usage

Flinit can be used on any system with Python and PIP installed and **PIP added to the OS Environment Path**.

## Installation

Flinit can be installed using Pip.

```
pip install flinit
```

## Using Flinit to Create a Flask Project

After installing flinit using pip, goto any CLI and use the <kbd>flinit</kbd> command to setup the Flask App.

```
flinit [projectName] [location] -g -r -c

flinit [projectName] [location] -gt -rd -cr

flinit [projectName] [location] --git --readme --cors
```

<kbd>projectName</kbd> - The Name of the Project.

<kbd>location</kbd>    - Path where Project is to be setup.

**Flags**

<kbd>-h</kbd> or <kbd>--help</kbd> - Provides Instruction about how to use Flinit.

<kbd>-g</kbd> or <kbd>-gt</kbd> or <kbd>--git</kbd> - Flag to add git to project. If this flag is used, git will be initialized for the project.

<kbd>-r</kbd> or <kbd>-rd</kbd> or <kbd>--readme</kbd> - Flag to add README.md to project. If this flag is used, a README file is generated.

<kbd>-c</kbd> or <kbd>-cr</kbd> or <kbd>--cors</kbd> - Flag to install and CORS to project. If this flag is used, CORS is installed and added a helper method is added to the app.py file.


# <kbd>flinit</kbd> in Action

**Flinit Helper**

!["Flinit Helper"](./examples/flinit-helper.png)

**Creating a Flask App <kbd>webapp</kbd>**

!["Flinit In Action"](./examples/flinit-in-action.png)

**Flask App webapp, initialized using <kbd>flinit</kbd>**

!["Flinit In Action"](./examples/file-structure.png)



[//]: https://stackabuse.com/executing-shell-commands-with-python/

[//]: https://stackoverflow.com/questions/1158076/implement-touch-using-python

[//]: https://youtu.be/BKvj4FH67H8?t=1296

[//]: https://dev.to/divshekhar/python-subprocess-execute-shell-commands-1bl2#:~:text=The%20subprocess.,object%20of%20CompletedProcess%20in%20Python.

[//]: https://code-maven.com/slides/python/argparse-named

[//]: https://stackoverflow.com/questions/53151899/python-how-to-script-virtual-environment-building-and-activation

CHANGELOG.md
# CHANGELOG


## **v0.2.4** <kbd>Latest</kbd>

- Using `virtualenv` instead of `venv` for Linux
- Fixes and Enhancements 


## **v0.2.3**

- Fixes and Enhancements 


## **v0.2.2**

- Fixes and Enhancements 

## **v0.2.1**

- Minor Release v0.2.1 
- Fixes and Enhancements 

## **v0.1.10** 

- Optimized gitignore creation.

## **v0.1.9** 

- Fixed Bugs - specific to flinit on macOS
- Added Support for using multiple variations of PIP in path. If pip is not added to the path it uses <kbd>python3 -m pip install</kbd>
```
Supports -

pip install
python3 -m pip install
```
- Code refactored to ensure separation OS specific tasks in workflow.

## v0.1.8
- Added Documentation
- Fixed Bugs
- Refactoring Code

## v0.1.7
- Added wider support
- Bug Fixes

## v0.1.6
- Bug Fixes

## v0.1.5
- Bug Fixes for windows

## v0.1.4
- Added README feature

## v0.1.3
- Added CORS feature
- Bug Fixes

## v0.1.2 
- Added GIT feature
- Fixed Bugs

## v0.1.1 <kbd>Initial</kbd>
- Initial Release
- Added Working directory creation task


test.py
import argparse
import os
import sys
import subprocess
from pathlib import Path
import shutil
from emoji import emojize


gitIGNOREtemplate = '''
__pycache__/
*.py[cod]
*$py.class
*.so
.Python
build/
develop-eggs/
dist/
downloads/
eggs/
.eggs/
lib/
lib64/
parts/
sdist/
var/
wheels/
pip-wheel-metadata/
share/python-wheels/
*.egg-info/
.installed.cfg
*.egg
MANIFEST
*.manifest
*.spec
pip-log.txt
pip-delete-this-directory.txt
htmlcov/
.tox/
.nox/
.coverage
.coverage.*
.cache
nosetests.xml
coverage.xml
*.cover
*.py,cover
.hypothesis/
.pytest_cache/
*.mo
*.pot
*.log
local_settings.py
db.sqlite3
db.sqlite3-journal
instance/
.webassets-cache
.scrapy
docs/_build/
target/
.ipynb_checkpoints
profile_default/
ipython_config.py
.python-version
__pypackages__/
celerybeat-schedule
celerybeat.pid
*.sage.py
.env
.venv
env/
venv/
ENV/
env.bak/
venv.bak/
.spyderproject
.spyproject
.ropeproject
/site
.mypy_cache/
.dmypy.json
dmypy.json
.pyre/
'''

APPpy_CALLER = {
    "flask": "app = flask.Flask(__name__)",
    "cors": "flask_cors.CORS(app)"
}

APPpy_TEMPLATE = '''\n\n@app.route("/") \ndef hello_world(): \n\treturn "<p>Hello, World!</p>"'''

# Convert Path to Work with present working system
def systemSpecificPath(workingPath):
    if sys.platform in ["cygwin", "win32"]:
        return workingPath.replace("/", "\\")
    elif sys.platform in ["linux",  "darwin"]:
        return workingPath.replace("\\", "/").replace("//", "/")

def getPyPreRequisites():
    pyPath = sys.executable
    pipPath = ""

    if shutil.which("pip") != None :
        pipPath = "pip"
    elif os.popen("python -m pip --version").read() !=None:
        pipPath = pyPath+" -m pip"
    else:
        os.popen(pyPath+" -m ensurepip --upgrade")
        if shutil.which("pip") != None :
            pipPath = "pip"
        elif os.popen("python -m pip --version").read() !=None:
            pipPath = pyPath+" -m pip"

    return pyPath, pipPath


def createProjectDirectory(workingDirectory, projectName):
    destination = systemSpecificPath(workingDirectory+"/"+projectName+"/")
    if not os.path.isdir(destination):
        os.system("mkdir {0}".format(destination))
        print(emojize(":check_mark_button:"), "Working Directory Created!")
    else:
        print(emojize(":check_mark_button:"), "Working Directory Exists!")
    return destination


def createAPPpy(workingDirectory):
    Path(workingDirectory+"/app.py").touch()
    print(emojize(":check_mark_button:"), "Created app.py!")


def importModuleToAPPpy(workingDirectory, package):
    with open(workingDirectory+"/app.py", "a+") as app:
        app.write("import "+package+"\n")


def addModuleCallerToAPPpy(workingDirectory, caller):
    with open(workingDirectory+"/app.py", "a+") as app:
        app.write(caller+"\n")


def writeTEMPLATEtoAPPpy(workingDirectory):
    with open(workingDirectory+"/app.py", "a+") as app:
        app.write(APPpy_TEMPLATE)


def createREADME(location, projectName):
    Path(location+"README.md").touch()
    with open(location+"README.md", "w+") as readme:
        readme.write("# {0} is a Python-Flask Project".format(projectName))
    print(emojize(":check_mark_button:"), "README.md Created!")


def generateGITignore(workingDirectory):
    Path(workingDirectory+".gitignore").touch()
    with open(workingDirectory+".gitignore", "a+") as gitignore:
        gitignore.write(gitIGNOREtemplate)
    print(emojize(":check_mark_button:"), ".gitignore Generated!")


def addGIT(workingDirectory):
    if shutil.which("git") != None:
        generateGITignore(workingDirectory)
        os.system("cd {0} && git init -q && git add . && git commit -q -m 'Initial'".format(
            systemSpecificPath(workingDirectory)))
        print(emojize(":check_mark_button:"), "Git Setup Complete.")
        print(emojize(":check_mark_button:"),
              "Code Initialized and Committed.")
    else:
        print(emojize(":cross_mark:"),
              "Git not Installed. Skipped setting up Git.")


class linuxUNIXRunner:
    def __init__(self, location, projectName, iCors=False, iREADME=False, iGit=False) -> None:
        print(emojize(":star-struck:"), "\033[93mWelcome to Flinit @ 0.1-beta\033[0m")
        self.pythonCommand , self.pipCommand = getPyPreRequisites()
        self.projectName = projectName
        self.location = location
        self.iCors = iCors
        self.iREADME = iREADME
        self.iGit = iGit
        self.workingDirectory = createProjectDirectory(self.location, self.projectName)

    def createVirtualEnvironment(self):
        os.system('cd {0} && {1} -m venv {2}'.format(
                systemSpecificPath(self.workingDirectory),
                self.pythonCommand,
                systemSpecificPath("virtualenv")
            )
        )
        print(emojize(":check_mark_button:"), "Virtual Environment Created!")

    def generateRequirementTXT(self):
        os.system('. {0} && {1} -q freeze > {2}/requirements.txt  -q'.format(
                systemSpecificPath(self.workingDirectory +"virtualenv/bin/activate"),
                self.pipCommand, 
                systemSpecificPath(self.workingDirectory)
            )
        )
        print(emojize(":check_mark_button:"), "Requirement.txt Generated!")

    def installFlask(self):
        os.system('. {0} && {1} -q install -q Flask'.format(
                systemSpecificPath(self.workingDirectory +"virtualenv/bin/activate"),
                self.pipCommand
            )
        )
        print(emojize(":check_mark_button:"), "Installed Flask!")

    def installCORS(self):
        os.system('. {0} && {1} -q install -U -q flask-cors'.format(
                systemSpecificPath(self.workingDirectory+"virtualenv/bin/activate"),
                self.pipCommand
            )
        )
        print(emojize(":check_mark_button:"), "Installed CORS!")

    def linuxUNIXFlowRunner(self):
        self.createVirtualEnvironment()
        createAPPpy(self.workingDirectory)
        self.installFlask()
        importModuleToAPPpy(self.workingDirectory, "flask")
        if self.iCors:
            self.installCORS()
            importModuleToAPPpy(self.workingDirectory, "flask_cors")
        addModuleCallerToAPPpy(self.workingDirectory, APPpy_CALLER["flask"])
        if self.iCors:
            addModuleCallerToAPPpy(self.workingDirectory, APPpy_CALLER["cors"])
        writeTEMPLATEtoAPPpy(self.workingDirectory)
        self.generateRequirementTXT()
        if self.iREADME:
            createREADME(self.workingDirectory, self.projectName)
        if self.iGit:
            addGIT(self.workingDirectory)
        print(emojize(":clinking_beer_mugs:"),
            "\033[93m\033[1mflinit Complete!\033[0m\033[0m")


class windowsRunner:
    def __init__(self, location, projectName, iCors=False, iREADME=False, iGit=False) -> None:
        print(emojize(":star-struck:"), "\033[93mWelcome to Flinit @ 0.1-beta\033[0m")
        self.pythonCommand , self.pipCommand = getPyPreRequisites()
        self.projectName = projectName
        self.location = location
        self.iCors = iCors
        self.iREADME = iREADME
        self.iGit = iGit
        self.workingDirectory = createProjectDirectory(self.location, self.projectName)

    def createVirtualEnvironment(self):
        os.system('cd {0} && {1} -m venv {2}'.format(
                systemSpecificPath(self.workingDirectory),
                self.pythonCommand,
                systemSpecificPath("virtualenv")
            )
        )
        print(emojize(":check_mark_button:"), "Virtual Environment Created!")

    def generateRequirementTXT(self):
        os.system('{0} && {1} -q freeze > {2}/requirements.txt  -q'.format(
                systemSpecificPath(self.workingDirectory+"\\virtualenv\\Scripts\\activate"),
                self.pipCommand,
                systemSpecificPath(self.workingDirectory)
            )
        )
        print(emojize(":check_mark_button:"), "Requirement.txt Generated!")

    def installFlask(self):
        os.system('{0} && {1} -q install -q Flask'.format(
                systemSpecificPath(self.workingDirectory+"\\virtualenv\\Scripts\\activate"),
                self.pipCommand
            )
        )
        print(emojize(":check_mark_button:"), "Installed Flask!")

    def installCORS(self):
        os.system('{0} && {1} -q install -U -q flask-cors'.format(
                systemSpecificPath(self.workingDirectory+"\\virtualenv\\Scripts\\activate"),
                self.pipCommand
            )
        )
        print(emojize(":check_mark_button:"), "Installed CORS!")

    def windowsFlowRunner(self):
        self.createVirtualEnvironment()
        createAPPpy(self.workingDirectory)
        self.installFlask()
        importModuleToAPPpy(self.workingDirectory, "flask")
        if self.iCors:
            self.installCORS()
            importModuleToAPPpy(self.workingDirectory, "flask_cors")
        addModuleCallerToAPPpy(self.workingDirectory, APPpy_CALLER["flask"])
        if self.iCors:
            addModuleCallerToAPPpy(self.workingDirectory, APPpy_CALLER["cors"])
        writeTEMPLATEtoAPPpy(self.workingDirectory)
        self.generateRequirementTXT()
        if self.iREADME:
            createREADME(self.workingDirectory, self.projectName)
        if self.iGit:
            addGIT(self.workingDirectory)
        print(emojize(":clinking_beer_mugs:"),
            "\033[93m\033[1mflinit Complete!\033[0m\033[0m")


def main():
    parser = argparse.ArgumentParser()

    parser.add_argument("projectName", type=str,
                        help="Name of the Project")

    parser.add_argument(
        "location", type=str, help="The Location where the Project is to be created")

    parser.add_argument("--git", "-gt", "-g", dest="needGIT", action="store_true",
                        help="Initialize Git for the Project. Works only if Git is installed.")

    parser.add_argument("--cors", "-cr", "-c", dest="needCORS", action="store_true",
                        help="Add Cross-origin resource sharing (CORS) to Flask App.")

    parser.add_argument("--readme", "-rd", "-r", dest="needREADME",
                        action="store_true", help="Add README.md for Project.")

    parser.add_argument('--version', "-v", action='version', version='flinit 0.1.9') # Upgrade Version on Every Publish

    parser.set_defaults(needGIT=False, needCORS=False, needREADME=False)

    args = parser.parse_args()

    if sys.platform in ["cygwin", "win32"]:
        runner = windowsRunner(
            location=args.location, 
            projectName=args.projectName,
            iCors=args.needCORS, 
            iREADME=args.needREADME, 
            iGit=args.needGIT
        )
        runner.windowsFlowRunner()
    elif sys.platform in ["linux",  "darwin"]:
        runner = linuxUNIXRunner(
            location=args.location,
            projectName=args.projectName,
            iCors=args.needCORS,
            iREADME=args.needREADME, 
            iGit=args.needGIT
        )
        runner.linuxUNIXFlowRunner()

test2.py
import argparse
import os
import sys
import subprocess
from pathlib import Path
import shutil
from emoji import emojize



def getPyPreRequisites():
    pyPath = "python"
    pipPath = "pip"

    if shutil.which("python") != None :
        if int(os.popen('python --version').read().split(" ")[1].replace("\n", "").split(".")[0]) >= 3 :
            pyPath = "python"
        else :
            print("Python 2 or lower is not supported")
    elif shutil.which("python3") != None :
        if int(os.popen('python3 --version').read().split(" ")[1].replace("\n", "").split(".")[0]) >= 3 :
            pyPath = "python3"
        else :
            print("Python 2 or lower is not supported")
    elif shutil.which("py") != None :
        if int(os.popen('py --version').read().split(" ")[1].replace("\n", "").split(".")[0]) >= 3 :
            pyPath = "py"
        else :
            print("Python 2 or lower is not supported")
    else :
        print("Python PATH issue")

    if shutil.which("pip") != None :
        pipPath = "pip"
    elif os.popen("python -m pip --version").read() !=None:
        pipPath = pyPath+" -m pip"
    else:
        os.popen(pyPath+" -m ensurepip --upgrade")
        if shutil.which("pip") != None :
            pipPath = "pip"
        elif os.popen("python -m pip --version").read() !=None:
            pipPath = pyPath+" -m pip"

    return pyPath, pipPath

    


def main():
    pythonCMD, pipCMD = getPyPreRequisites()
    # print(os.popen("python -m pip --version").read())
    print(pythonCMD, pipCMD)


if __name__ == "__main__":
    main()

flinit-Backup.py
import argparse
import os
import sys
import subprocess
from pathlib import Path
import shutil
from emoji import emojize


gitIGNOREtemplate = '''
__pycache__/
*.py[cod]
*$py.class
*.so
.Python
build/
develop-eggs/
dist/
downloads/
eggs/
.eggs/
lib/
lib64/
parts/
sdist/
var/
wheels/
pip-wheel-metadata/
share/python-wheels/
*.egg-info/
.installed.cfg
*.egg
MANIFEST
*.manifest
*.spec
pip-log.txt
pip-delete-this-directory.txt
htmlcov/
.tox/
.nox/
.coverage
.coverage.*
.cache
nosetests.xml
coverage.xml
*.cover
*.py,cover
.hypothesis/
.pytest_cache/
*.mo
*.pot
*.log
local_settings.py
db.sqlite3
db.sqlite3-journal
instance/
.webassets-cache
.scrapy
docs/_build/
target/
.ipynb_checkpoints
profile_default/
ipython_config.py
.python-version
__pypackages__/
celerybeat-schedule
celerybeat.pid
*.sage.py
.env
.venv
env/
venv/
ENV/
env.bak/
venv.bak/
.spyderproject
.spyproject
.ropeproject
/site
.mypy_cache/
.dmypy.json
dmypy.json
.pyre/
'''

APPpy_CALLER = {
    "flask": "app = flask.Flask(__name__)",
    "cors": "flask_cors.CORS(app)"
}

APPpy_TEMPLATE = '''\n@app.route("/") \ndef hello_world(): \n\treturn "<p>Hello, World!</p>"'''


def systemSpecificPath(workingPath):
    if sys.platform in ["cygwin", "win32"]:
        return workingPath.replace("/", "\\")
    elif sys.platform in ["linux",  "darwin"]:
        return workingPath.replace("\\", "/").replace("//", "/")


def createAPPpy(workingDirectory):
    Path(workingDirectory+"/app.py").touch()
    print(emojize(":check_mark_button:"), "Created app.py!")


def importModuleToAPPpy(workingDirectory, package):
    with open(workingDirectory+"/app.py", "a+") as app:
        app.write("import "+package+"\n")


def addModuleCallerToAPPpy(workingDirectory, caller):
    with open(workingDirectory+"/app.py", "a+") as app:
        app.write(caller+"\n")


def writeTEMPLATEtoAPPpy(workingDirectory):
    with open(workingDirectory+"/app.py", "a+") as app:
        app.write(APPpy_TEMPLATE)


def generateRequirementTXT(workingDirectory):
    if sys.platform in ["cygwin", "win32"]:
        os.system('{0} && pip -q freeze > {1}/requirements.txt  -q'.format(systemSpecificPath(
            workingDirectory+"\\virtualenv\\Scripts\\activate"), systemSpecificPath(workingDirectory)))
    elif sys.platform in ["linux",  "darwin"]:
        os.system('. {0} && pip -q freeze > {1}/requirements.txt  -q'.format(systemSpecificPath(
            workingDirectory+"virtualenv/bin/activate"), systemSpecificPath(workingDirectory)))
    print(emojize(":check_mark_button:"), "Requirement.txt Generated!")


def installFlask(workingDirectory):
    if sys.platform in ["cygwin", "win32"]:
        os.system('{0} && pip -q install -q Flask'.format(systemSpecificPath(
            workingDirectory+"\\virtualenv\\Scripts\\activate")))
    elif sys.platform in ["linux",  "darwin"]:
        os.system('. {0} && pip -q install -q Flask'.format(
            systemSpecificPath(workingDirectory+"virtualenv/bin/activate")))
    print(emojize(":check_mark_button:"), "Installed Flask!")


def installCORS(workingDirectory):
    if sys.platform in ["cygwin", "win32"]:
        os.system('{0} && pip -q install -U -q flask-cors'.format(systemSpecificPath(
            workingDirectory+"\\virtualenv\\Scripts\\activate")))
    elif sys.platform in ["linux",  "darwin"]:
        os.system('. {0} && pip -q install -U -q flask-cors'.format(
            systemSpecificPath(workingDirectory+"virtualenv/bin/activate")))
    print(emojize(":check_mark_button:"), "Installed CORS!")


def createProjectDirectory(workingDirectory, projectName):
    destination = systemSpecificPath(workingDirectory+"/"+projectName+"/")
    if not os.path.isdir(destination):
        os.system("mkdir {0}".format(destination))
        print(emojize(":check_mark_button:"), "Working Directory Created!")
    else:
        print(emojize(":check_mark_button:"), "Working Directory Exists!")
    return destination


def createVirtualEnvironment(workingDirectory):
    if sys.platform != "aix":
        try:
            os.system("python -m venv " +
                      systemSpecificPath(workingDirectory+"virtualenv"))
            print(emojize(":check_mark_button:"),
                  "Virtual Environment Created!")
        except subprocess.CalledProcessError as e:
            print(
                "There was an error creating virtual environment. Check the error in errors.txt")
            with open(workingDirectory+"/error.txt", "w+") as err:
                err.write(e)
            print(emojize(":cross_mark:"), "Flinit Initialize Project Failed")
            sys.exit()
    else:
        print(emojize(":cross_mark:"),
              "System not supported. Flinit Initialize Project Failed")
        sys.exit()


def createREADME(location, projectName):
    Path(location+"README.md").touch()
    with open(location+"README.md", "w+") as readme:
        readme.write("# {0} is a Python-Flask Project".format(projectName))
    print(emojize(":check_mark_button:"), "README.md Created!")


def generateGITignore(workingDirectory):
    Path(workingDirectory+".gitignore").touch()
    with open(workingDirectory+".gitignore", "a+") as gitignore:
        gitignore.write(gitIGNOREtemplate)
    print(emojize(":check_mark_button:"), ".gitignore Generated!")


def addGIT(workingDirectory):
    if shutil.which("git") != None:
        generateGITignore(workingDirectory)
        os.system("cd {0} && git init -q && git add . && git commit -q -m 'Initial'".format(
            systemSpecificPath(workingDirectory)))
        print(emojize(":check_mark_button:"), "Git Setup Complete.")
        print(emojize(":check_mark_button:"),
              "Code Initialized and Committed.")
    else:
        print(emojize(":cross_mark:"),
              "Git not Installed. Skipped setting up Git.")


def runner(location, projectName, iCors=False, iREADME=False, iGit=False):
    print(emojize(":star-struck:"),
          "\033[93mWelcome to Flinit @ 0.1-beta\033[0m")
    workingDirectory = createProjectDirectory(location, projectName)
    createVirtualEnvironment(workingDirectory)
    createAPPpy(workingDirectory)
    installFlask(workingDirectory)
    importModuleToAPPpy(workingDirectory, "flask")
    if iCors:
        installCORS(workingDirectory)
        importModuleToAPPpy(workingDirectory, "flask_cors")
    addModuleCallerToAPPpy(workingDirectory, APPpy_CALLER["flask"])
    if iCors:
        addModuleCallerToAPPpy(workingDirectory, APPpy_CALLER["cors"])
    writeTEMPLATEtoAPPpy(workingDirectory)
    generateRequirementTXT(workingDirectory)
    if iREADME:
        createREADME(workingDirectory, projectName)
    if iGit:
        addGIT(workingDirectory)
    print(emojize(":clinking_beer_mugs:"),
          "\033[93m\033[1mflinit Complete!\033[0m\033[0m")


def main():
    parser = argparse.ArgumentParser()

    parser.add_argument("projectName", type=str,
                        help="Name of the Project")

    parser.add_argument(
        "location", type=str, help="The Location where the Project is to be created")

    parser.add_argument("--git", "-gt", "-g", dest="needGIT", action="store_true",
                        help="Initialize Git for the Project. Works only if Git is installed.")

    parser.add_argument("--cors", "-cr", "-c", dest="needCORS", action="store_true",
                        help="Add Cross-origin resource sharing (CORS) to Flask App.")

    parser.add_argument("--readme", "-rd", "-r", dest="needREADME",
                        action="store_true", help="Add README.md for Project.")

    parser.set_defaults(needGIT=False, needCORS=False, needREADME=False)

    args = parser.parse_args()

    runner(location=args.location, projectName=args.projectName,
           iCors=args.needCORS, iREADME=args.needREADME, iGit=args.needGIT)


if __name__ == "__main__":
    main()

flinit.py
import argparse
import os
import sys
from pathlib import Path
import shutil
from emoji import emojize

gitIGNOREtemplateURL = "https://raw.githubusercontent.com/github/gitignore/master/Python.gitignore"

APPpy_CALLER = {
    "flask": "app = flask.Flask(__name__)",
    "cors": "flask_cors.CORS(app)"
}

APPpy_TEMPLATE = '''\n\n@app.route("/") \ndef hello_world(): \n\treturn "<p>Hello, World!</p>"'''

# Convert Path to Work with present working system


def systemSpecificPath(workingPath):
    if sys.platform in ["cygwin", "win32"]:
        return workingPath.replace("/", "\\")
    elif sys.platform in ["linux",  "darwin"]:
        return workingPath.replace("\\", "/").replace("//", "/")


def getPyPreRequisites():
    pyPath = sys.executable
    pipPath = ""

    if shutil.which("pip") != None:
        pipPath = "pip"
    elif os.popen("python -m pip --version").read() != None:
        pipPath = pyPath+" -m pip"
    else:
        os.popen(pyPath+" -m ensurepip --upgrade")
        if shutil.which("pip") != None:
            pipPath = "pip"
        elif os.popen("python -m pip --version").read() != None:
            pipPath = pyPath+" -m pip"
    return pyPath, pipPath


def createProjectDirectory(workingDirectory, projectName):
    destination = systemSpecificPath(workingDirectory+"/"+projectName+"/")
    if not os.path.isdir(destination):
        os.system("mkdir {0}".format(destination))
        print(emojize(":check_mark_button:"), "Working Directory Created!")
    else:
        print(emojize(":check_mark_button:"), "Working Directory Exists!")
    return destination


def createAPPpy(workingDirectory):
    Path(workingDirectory+"/app.py").touch()
    print(emojize(":check_mark_button:"), "Created app.py!")


def importModuleToAPPpy(workingDirectory, package):
    with open(workingDirectory+"/app.py", "a+") as app:
        app.write("import "+package+"\n")


def addModuleCallerToAPPpy(workingDirectory, caller):
    with open(workingDirectory+"/app.py", "a+") as app:
        app.write(caller+"\n")


def writeTEMPLATEtoAPPpy(workingDirectory):
    with open(workingDirectory+"/app.py", "a+") as app:
        app.write(APPpy_TEMPLATE)


def createREADME(location, projectName):
    Path(location+"README.md").touch()
    with open(location+"README.md", "w+") as readme:
        readme.write("# {0} is a Python-Flask Project".format(projectName))
    print(emojize(":check_mark_button:"), "README.md Created!")


def generateGITignore(workingDirectory):
    os.system("cd {0} && curl -LJO -s {1}".format(
            systemSpecificPath(workingDirectory),
            gitIGNOREtemplateURL
        )
    )
    Path(workingDirectory+".gitignore").touch()
    shutil.copyfile("{0}/Python.gitignore".format(workingDirectory), "{0}/.gitignore".format(workingDirectory))
    with open("{0}/.gitignore".format(workingDirectory), "a") as dotGITIGNORE:
        dotGITIGNORE.write("\n# Virtual Environment\nvirtualenv/")
    os.remove("{0}/Python.gitignore".format(workingDirectory))


def addGIT(workingDirectory):
    if shutil.which("git") != None:
        generateGITignore(workingDirectory)
        os.system("cd {0} && git init -q".format(systemSpecificPath(workingDirectory)))
        # os.system("cd {0} && git add . ".format(systemSpecificPath(workingDirectory)))
        # os.system("cd {0} && git commit -q -m 'Initial Commit'".format(systemSpecificPath(workingDirectory)))
        print(emojize(":check_mark_button:"), "Git Initialized.")
        print(emojize(":check_mark_button:"), "Git Setup Complete.")
    else:
        print(emojize(":cross_mark:"),
              "Git not Installed. Skipped setting up Git.")


class linuxUNIXRunner:
    def __init__(self, location, projectName, iCors=False, iREADME=False, iGit=False) -> None:
        print(emojize(":star-struck:"),
              "\033[93mWelcome to Flinit @ 0.1-beta\033[0m")
        self.pythonCommand, self.pipCommand = getPyPreRequisites()
        self.projectName = projectName
        self.location = location
        self.iCors = iCors
        self.iREADME = iREADME
        self.iGit = iGit
        self.workingDirectory = createProjectDirectory(
            self.location, self.projectName)

    def createVirtualEnvironment(self):
        os.system('{0} -q install -q virtualenv'.format(self.pipCommand))
        os.system('cd {0} && virtualenv {1} -q'.format(
                systemSpecificPath(self.workingDirectory),
                systemSpecificPath("virtualenv")
            )
        )
        print(emojize(":check_mark_button:"), "Virtual Environment Created!")

    def generateRequirementTXT(self):
        os.system('. {0} && {1} -q freeze > {2}/requirements.txt  -q'.format(
            systemSpecificPath(self.workingDirectory +
                               "virtualenv/bin/activate"),
            self.pipCommand,
            systemSpecificPath(self.workingDirectory)
        )
        )
        print(emojize(":check_mark_button:"), "Requirement.txt Generated!")

    def installFlask(self):
        os.system('. {0} && {1} -q install -q Flask'.format(
                systemSpecificPath(self.workingDirectory +"virtualenv/bin/activate"),
                self.pipCommand
            )
        )
        print(emojize(":check_mark_button:"), "Installed Flask!")

    def installCORS(self):
        os.system('. {0} && {1} -q install -U -q flask-cors'.format(
            systemSpecificPath(self.workingDirectory +
                               "virtualenv/bin/activate"),
            self.pipCommand
        )
        )
        print(emojize(":check_mark_button:"), "Installed CORS!")

    def linuxUNIXFlowRunner(self):
        self.createVirtualEnvironment()
        createAPPpy(self.workingDirectory)
        self.installFlask()
        importModuleToAPPpy(self.workingDirectory, "flask")
        if self.iCors:
            self.installCORS()
            importModuleToAPPpy(self.workingDirectory, "flask_cors")
        addModuleCallerToAPPpy(self.workingDirectory, APPpy_CALLER["flask"])
        if self.iCors:
            addModuleCallerToAPPpy(self.workingDirectory, APPpy_CALLER["cors"])
        writeTEMPLATEtoAPPpy(self.workingDirectory)
        self.generateRequirementTXT()
        if self.iREADME:
            createREADME(self.workingDirectory, self.projectName)
        if self.iGit:
            addGIT(self.workingDirectory)
        print(emojize(":clinking_beer_mugs:"),
              "\033[93m\033[1mflinit Complete!\033[0m\033[0m")


class windowsRunner:
    def __init__(self, location, projectName, iCors=False, iREADME=False, iGit=False) -> None:
        print(emojize(":star-struck:"),
              "\033[93mWelcome to Flinit @ 0.1-beta\033[0m")
        self.pythonCommand, self.pipCommand = getPyPreRequisites()
        self.projectName = projectName
        self.location = location
        self.iCors = iCors
        self.iREADME = iREADME
        self.iGit = iGit
        self.workingDirectory = createProjectDirectory(
            self.location, self.projectName)

    def createVirtualEnvironment(self):
        os.system('cd {0} && {1} -m venv {2}'.format(
            systemSpecificPath(self.workingDirectory),
            self.pythonCommand,
            systemSpecificPath("virtualenv")
        )
        )
        print(emojize(":check_mark_button:"), "Virtual Environment Created!")

    def generateRequirementTXT(self):
        os.system('{0} && {1} -q freeze > {2}/requirements.txt  -q'.format(
            systemSpecificPath(self.workingDirectory +
                               "\\virtualenv\\Scripts\\activate"),
            self.pipCommand,
            systemSpecificPath(self.workingDirectory)
        )
        )
        print(emojize(":check_mark_button:"), "Requirement.txt Generated!")

    def installFlask(self):
        os.system('{0} && {1} -q install -q Flask'.format(
            systemSpecificPath(self.workingDirectory +
                               "\\virtualenv\\Scripts\\activate"),
            self.pipCommand
        )
        )
        print(emojize(":check_mark_button:"), "Installed Flask!")

    def installCORS(self):
        os.system('{0} && {1} -q install -U -q flask-cors'.format(
            systemSpecificPath(self.workingDirectory +
                               "\\virtualenv\\Scripts\\activate"),
            self.pipCommand
        )
        )
        print(emojize(":check_mark_button:"), "Installed CORS!")

    def windowsFlowRunner(self):
        self.createVirtualEnvironment()
        createAPPpy(self.workingDirectory)
        self.installFlask()
        importModuleToAPPpy(self.workingDirectory, "flask")
        if self.iCors:
            self.installCORS()
            importModuleToAPPpy(self.workingDirectory, "flask_cors")
        addModuleCallerToAPPpy(self.workingDirectory, APPpy_CALLER["flask"])
        if self.iCors:
            addModuleCallerToAPPpy(self.workingDirectory, APPpy_CALLER["cors"])
        writeTEMPLATEtoAPPpy(self.workingDirectory)
        self.generateRequirementTXT()
        if self.iREADME:
            createREADME(self.workingDirectory, self.projectName)
        if self.iGit:
            addGIT(self.workingDirectory)
        print(emojize(":clinking_beer_mugs:"),
              "\033[93m\033[1mflinit Complete!\033[0m\033[0m")


def main():
    parser = argparse.ArgumentParser()

    parser.add_argument("projectName", type=str,
                        help="Name of the Project")

    parser.add_argument(
        "location", type=str, help="The Location where the Project is to be created")

    parser.add_argument('--version', "-v", action='version', version='flinit 0.2.3 Beta',
                        help="Version of flinit")  # Upgrade Version on Every Publish

    parser.add_argument("--git", "-gt", "-g", dest="needGIT", action="store_true",
                        help="Initialize Git for the Project. Works only if Git is installed.")

    parser.add_argument("--cors", "-cr", "-c", dest="needCORS", action="store_true",
                        help="Add Cross-origin resource sharing (CORS) to Flask App.")

    parser.add_argument("--readme", "-rd", "-r", dest="needREADME",
                        action="store_true", help="Add README.md for Project.")

    parser.set_defaults(needGIT=False, needCORS=False, needREADME=False)

    args = parser.parse_args()

    if sys.platform in ["cygwin", "win32"]:
        runner = windowsRunner(
            location=args.location,
            projectName=args.projectName,
            iCors=args.needCORS,
            iREADME=args.needREADME,
            iGit=args.needGIT
        )
        runner.windowsFlowRunner()
    elif sys.platform in ["linux",  "darwin"]:
        runner = linuxUNIXRunner(
            location=args.location,
            projectName=args.projectName,
            iCors=args.needCORS,
            iREADME=args.needREADME,
            iGit=args.needGIT
        )
        runner.linuxUNIXFlowRunner()

if __name__ == "__main__":
    main()